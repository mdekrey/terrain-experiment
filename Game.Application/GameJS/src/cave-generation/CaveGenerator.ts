import {
  GameCoordinates
} from "../game/GameCoordinates";
import { ajax } from "rxjs/ajax";

export interface Cave {
  offset: GameCoordinates;
  isSolid: boolean[][];
  treasure: GameCoordinates[];
  entrance: GameCoordinates;
}

export class CaveGenerator {
  private readonly result: Promise<{
    isSolid: boolean[][];
    entrance: GameCoordinates;
    treasure: GameCoordinates[];
  }>;
  private readonly position: GameCoordinates;
  private readonly offset: GameCoordinates;

  constructor(
    position: GameCoordinates,
    offset: GameCoordinates
  ) {
    this.position = position;
    this.offset = offset;

    this.result = new Promise(this.buildMap);
  }

  private readonly buildMap = async (
    resolve: (t: { isSolid: boolean[][]; treasure: GameCoordinates[], entrance: GameCoordinates }) => void
  ) => {
    const response = await ajax({ url: "/api/terrain/cave", body: { Coordinates: { X: Math.round(this.position.x), Y: Math.round(this.position.y) } }, method: "POST", headers: { 'Content-Type': 'application/json' } })
      .toPromise();
    const result = response.response as { map: boolean[][]; treasure: GameCoordinates[], entrance: GameCoordinates };
    resolve({ isSolid: result.map, treasure: result.treasure, entrance: result.entrance });
  };

  get cave(): Promise<Cave> {
    return this.result.then(cave => ({
      isSolid: cave.isSolid,
      entrance: cave.entrance,
      treasure: cave.treasure,
      offset: this.offset
    }));
  }
}
