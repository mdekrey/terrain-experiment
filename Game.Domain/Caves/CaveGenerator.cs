using LibNoise;
using LibNoise.Generator;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Game.Domain.Caves
{
    public class CaveGenerator
    {
        const int deathLimit = 3;
        const int birthLimit = 4;
        const int treasureLimit = 5;

        private readonly int width;
        private readonly int height;
        private readonly int normalizationRounds;
        private readonly ModuleBase perlin;
        private bool[][] map;
        private readonly Task<Cave> building;



        bool Random(int x, int y) => perlin.GetValue(x / 3f, y / 3f, 0) < 0;

        public CaveGenerator(int width, int height, int normalizationRounds, int seed)
        {
            this.width = width;
            this.height = height;
            this.normalizationRounds = normalizationRounds;

            this.perlin = new Perlin(lacunarity: 3.4f, seed: seed);
            this.map = (from y in Enumerable.Range(0, height)
                        select (from x in Enumerable.Range(0, width)
                                select Random(x, y)).ToArray()).ToArray();
            building = BuildMap();
        }

        private int CountAliveNeighbours(CaveCoordinate position)
        {
            var count = 0;
            foreach (var dir in CaveCoordinate.EightDirections)
            {
                var nb = dir + position;
                //If it's at the edges, consider it to be REALLY solid
                if (
                  nb.x < 0 ||
                  nb.y < 0 ||
                  nb.x >= width ||
                  nb.y >= height
                )
                {
                    count = count + 2;
                }
                else if (this.map[nb.y][nb.x])
                {
                    count = count + 1;
                }
            }
            return count;
        }

        private async Task<Cave> BuildMap()
        {
            await Task.Yield();
            for (var count = 0; count < this.normalizationRounds; count++)
            {
                this.DoSimulationStep();
            }

            this.FillSmallerAreas();
            this.RemoveSingleY();

            var treasure = FindTreasure();
            return new Cave
            {
                Map = map,
                Entrance = treasure[0],
                Treasure = treasure.Skip(1).ToArray(),
            };
        }

        private CaveCoordinate[] FindTreasure()
        {
            return (from y in this.map.Select((_, y) => y)
                    from v in this.map[y].Select((currentCell, x) => (currentCell, x))
                    let currentCell = v.currentCell
                    where !currentCell
                    let x = v.x
                    let r = new CaveCoordinate(x, y)
                    let nbs = this.CountAliveNeighbours(r)
                    where nbs >= treasureLimit
                    select r).ToArray();
        }

        private void FillSmallerAreas()
        {
            var areas = new Dictionary<string, int>();
            string[][] currentAreas = this.map.Select(_ => new string[_.Length]).ToArray();

            var threshold = (this.width * this.height) / 4;
            string shortcutSymbol = null;
            for (var y = 0; y < this.map.Length && shortcutSymbol == null; y++)
            {
                for (var x = 0; x < this.map[y].Length && shortcutSymbol == null; x++)
                {
                    if (!this.map[y][x] && currentAreas[y][x] == null)
                    {
                        var (symbol, count) = this.FloodFill(
                          new CaveCoordinate(x, y),
                          currentAreas,
                          areas
                        );
                        if (count > threshold)
                        {
                            shortcutSymbol = symbol;
                        }
                    }
                }
            }
            var topSymbol =
              shortcutSymbol ?? areas.OrderByDescending(v => v.Value).First().Key;

            for (var y = 0; y < this.map.Length; y++)
            {
                for (var x = 0; x < this.map[y].Length; x++)
                {
                    if (currentAreas[y][x] != topSymbol)
                    {
                        this.map[y][x] = true;
                    }
                }
            }
        }

        private (string symbol, int count) FloodFill(CaveCoordinate start, string[][] currentAreas, Dictionary<string, int> areas)
        {
            var symbol = $"{start.x}x{start.y}";
            var count = 0;
            var queue = new Queue<CaveCoordinate>();
            queue.Enqueue(start);
            while (queue.Count > 0)
            {
                var n = queue.Dequeue();
                foreach (var dir in CaveCoordinate.FourDirections)
                {
                    var r = (n + dir);
                    if (!this.map[r.y][r.x] && currentAreas[r.y][r.x] == null)
                    {
                        count++;
                        currentAreas[r.y][r.x] = symbol;
                        queue.Enqueue(r);
                    }
                }
            }
            areas[symbol] = count;
            return (symbol, count);
        }

        void RemoveSingleY()
        {
            for (var y = 1; y < this.map.Length - 1; y++)
            {
                for (var x = 0; x < this.map[y].Length; x++)
                {
                    if (
                      this.map[y][x] &&
                      !this.map[y + 1][x] &&
                      !this.map[y - 1][x]
                    )
                    {
                        this.map[y][x] = false;
                    }
                }
            }
        }

        private void DoSimulationStep()
        {
            this.map = this.map.Select((_, y) => _.Select((currentCell, x) =>
            {
                var nbs = CountAliveNeighbours(new CaveCoordinate(x, y));
                if (currentCell)
                {
                    return nbs >= deathLimit;
                }
                else
                {
                    return nbs > birthLimit;
                }
            }).ToArray()).ToArray();
        }

        public Task<Cave> Generate()
        {
            return building;
        }
    }
}
