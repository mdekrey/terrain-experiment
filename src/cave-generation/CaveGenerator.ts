import { Perlin } from "libnoise-ts/module/generator";

const deathLimit = 3;
const birthLimit = 4;

export class CaveGenerator {
    private readonly perlin = new Perlin();
    private readonly result: Promise<boolean[][]>;

    constructor(seed: number, width: number, height: number, normalizationRounds: number) {
        this.perlin.seed = seed;
        // this.perlin.frequency = 0.5;
        this.perlin.lacunarity = 3.4;

        const random = (x: number, y: number) => this.perlin.getValue(x, y, 0) < -0.05;
        // const random = (x: number, y: number) => Math.random() < 0.45;

        this.result = new Promise<boolean[][]>((resolve, reject) => {
            let map = Array.from(Array(height).keys()).map((_, y) => Array.from(Array(width).keys()).map((_, x) => random(x, y)));
            const v = map.reduce((prev, next) => [...prev, ...next], []);
            let count = 0;

            function countAliveNeighbours(map: boolean[][], x: number, y: number){
                let count = 0;
                for(let i = -1; i < 2; i++){
                    for(let j = -1; j < 2; j++){
                        const nb_x = i+x;
                        const nb_y = j+y;
                        if(i == 0 && j == 0){
                        }
                        //If it's at the edges, consider it to be REALLY solid
                        else if(nb_x < 0 || nb_y < 0 ||
                                nb_x >= map.length ||
                                nb_y >= map[0].length){
                            count = count + 2;
                        }
                        else if(map[nb_y][nb_x]){
                            count = count + 1;
                        }
                    }
                }
                return count;
            }

            function doSimulationStep() {
                map = map.map((_, y) => _.map((currentCell, x) => {
                    const nbs = countAliveNeighbours(map, x, y);
                    if (currentCell) {
                        return nbs >= deathLimit;
                    } else {
                        return nbs > birthLimit;
                    }
                }));
            }

            function step() {
                if (count++ < normalizationRounds) {
                    doSimulationStep();
                    setTimeout(step);
                } else {
                    resolve(map);
                }
            }

            setTimeout(step)
        });
    }

    get map() {
        return this.result;
    }
}
