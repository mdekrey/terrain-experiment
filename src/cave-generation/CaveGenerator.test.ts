import { CaveGenerator } from "./CaveGenerator";

it("generates some terrain", async () => {
    const cave = new CaveGenerator(0, 50, 50, 2);

    const map = await cave.map;
    const mapDisplay = map.map(row => row.map(e => e ? "â–ˆ" : " ").join("")).join("\n");

    expect(mapDisplay).toMatchSnapshot();
});
