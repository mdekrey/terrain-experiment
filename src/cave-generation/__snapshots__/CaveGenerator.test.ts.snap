// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates some terrain 1`] = `
"██████████████████████████████████████████████████
█████████████  █████    █████   ██████████████████
████████████   ████      ████   █████   ██████████
█  ███   ████ ████    █ ███████████    ███████  ██
██ █      ███████       ██████████    ███████   ██
███       ███████        █    █       ███ ██    ██
███       ███████                          ██   ██
███      ████  ██        ███              ██     █
██      ████████         ████   ██        ██    ██
██     █████████         ████   ██   ██   ██  ████
███    ███  ████   ████  ████   ███       ████████
███    ██    ███  █████  ███     ██       ████████
██ █   ███   ████  ███   ███               ███████
█  ██   ██████████████  ████     ██        ███████
█  ██     █████████████████   ██████          ████
█  █      ███████████████    ███████           ███
███       ██████████████  ██████████           ███
███       ███████   ████████████████       ██   ██
██            ██    ████████████████       ██    █
█                    ███████████████   ██   █    █
█              ████   ██████   █  ██   ██   ██   █
█             █████                   ███   ██  ██
█             ████                    ██   ██    █
█              ████                        ██    █
█  ██  █  ██  ████████      ███             ███  █
███████████████████████    █████████     █   █████
████████████████████████████████████   ████  █████
███████████          ██████   ███████ █████   ████
██████████           █████       ████ ███       ██
██   ████         █████████      ██             ██
█     ███  ██ █████████████     ██            ████
█     ██  ████  ███████████     █             ████
█        █████  ██  ███████    █   █████  ████████
█        ████        ██████   ███ █████████████  █
█        ████           ████ ███████████████ ██  █
█        ████           ████  █████████████   █  █
█        █████ ██      █████      ████████      ██
█  ██ ██  ███████     ██████       ███████     ███
███  ███   ██████     ██  ██        ██████      ██
███   █    █████          ███                   ██
██         █████         █████                 ███
██    █   █████      █████████                ████
███████   █████      ██████████             ██████
██████    ██  █      ████ ██████            ██████
███                    ██  █████          ██    ██
█                 █     █  █████   █    ███     ██
█                ███     ███████        ███     ██
██████        ██████     █████████      ██       █
█████████    ████████    ███████████   ███      ██
██████████████████████████████████████████████████"
`;
